File: daily_problems/2025/10/1024/personal_submission/cf106129f_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!
// I came, I divided, I conquered!
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n, m;
	cin >> n >> m;

	vector<string> grid(n);
	for (auto &s: grid) cin >> s;

	int total = 0;
	for (auto &s: grid) {
		for (auto &c: s) {
			total += (c == '#');
		}
	}

	vector<vector<int>> tmp(n, vector<int>(m, 0));
	for (int i = min(n, m); i > 0; i --) {
		if (total % (i * i) == 0) {

			for (int x = 0; x < n; x ++) {
				for (int y = 0; y < m; y ++) {
					if (grid[x][y] == '#') {
						tmp[x][y] = 1;
					}
				}
			}

			bool flg = true;
			for (int x = 0; x < n; x ++) {
				for (int y = 0; y < m; y ++) {
					if (tmp[x][y]) {
						for (int dx = x; dx < x + i; dx ++) {
							for (int dy = y; dy < y + i; dy ++) {
								if (dx >= n || dy >= m || !tmp[dx][dy]) {
									flg = false;
									break;
								}
								tmp[dx][dy] = 0;
							}
							if (!flg) break;
						}
						if (!flg) break;
					}
				}
				if (!flg) break;
			}

			if (flg) return cout << i, 0;
		}
	}

	return 0;
}

File: daily_problems/2025/10/1024/personal_submission/cf106129f_yawn_sean.py
# Submission link: https://codeforces.com/gym/106129/submission/345310484
def main(): 
    n, m = MII()
    grid = [[1 if c == '#' else 0 for c in I()] for _ in range(n)]
    total = sum(sum(x) for x in grid)
    
    tmp = [[0] * m for _ in range(n)]
    
    for i in range(fmin(n, m), 0, -1):
        if total % (i * i) == 0:
            for x in range(n):
                for y in range(m):
                    tmp[x][y] = grid[x][y]
            
            flg = True
            for x in range(n):
                for y in range(m):
                    if tmp[x][y]:
                        for dx in range(x, x + i):
                            for dy in range(y, y + i):
                                if dx >= n or dy >= m or tmp[dx][dy] == 0:
                                    flg = False
                                    break
                                tmp[dx][dy] = 0
                            if not flg: break
                        if not flg: break
                if not flg: break
            
            if flg:
                print(i)
                break